src/scheduler.rs

```diff
@@ -22,6 +22,34 @@ pub struct Scheduler<T, W> {
     nodes: List<T, W>,
 }
 
+/// Push new futures to be scheduled.
+///
+/// This is used as a staging area for futures when either `&mut` access to
+/// `Scheduler` is not available due to borrow checker constraints *or* wanting
+/// to store push futures into a `Scheduler` via a thread-local.
+///
+/// The reason a `Scheduler` cannot be used directly from a thread local is due
+/// to the `W` generic. In order to set a scheduler to a thread local, the `W`
+/// generic must be set at compile time. This prevents letting the user run a
+/// scheduler with a custom wake up generic (for example, a Tokio reactor).
+///
+/// `Push` sets the node's `W` generic to `Void`, a type that cannot be
+/// instantiated. Then, when the node is added to a scheduler, it is transmuted
+/// to the correct `W` type.
+///
+/// This is safe for the following reasons
+///
+/// 1) The only usage of `W` in `Node` is `Option<Arc<Inner<_, W>>>`. When used
```

Comment:
I feel pretty uneasy about this approach; it's "probably fine" but we don't have enough representation guarantees to be sure.

I'd be a lot more comfortable using a `SmallVec` or something similar to queue up tasks, creating nodes afterward; assuming that most polling will result in at most a couple of new tasks, this is likely to be extremely small overhead and can be done in safe code.